samples/bluetooth/central_iso/src/main.c

```diff
@@ -19,64 +19,75 @@
 #include <zephyr/settings/settings.h>
 #include <zephyr/sys/byteorder.h>
 
+#define LATENCY_MS               10U                   /* 10.0 ms */
+#define INTERVAL_US              (10U * USEC_PER_MSEC) /* 10.0 ms */
+
 static void start_scan(void);
 
 static struct bt_conn *default_conn;
-static struct k_work_delayable iso_send_work;
 static struct bt_iso_chan iso_chan;
-static uint16_t seq_num;
-static uint16_t latency_ms = 10U; /* 10ms */
-static uint32_t interval_us = 10U * USEC_PER_MSEC; /* 10 ms */
 NET_BUF_POOL_FIXED_DEFINE(tx_pool, 1, BT_ISO_SDU_BUF_SIZE(CONFIG_BT_ISO_TX_MTU),
 			  CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
 
 /**
- * @brief Send ISO data on timeout
- *
- * This will send an increasing amount of ISO data, starting from 1 octet.
- *
- * First iteration : 0x00
- * Second iteration: 0x00 0x01
- * Third iteration : 0x00 0x01 0x02
- *
- * And so on, until it wraps around the configured ISO TX MTU (CONFIG_BT_ISO_TX_MTU)
+ * @brief Send ISO data SDU
  *
- * @param work Pointer to the work structure
+ * This will send a decreasing amount of ISO data until it reaches 1 byte then
+ * it wraps around.
  */
-static void iso_timer_timeout(struct k_work *work)
+static void iso_send_sdu(uint32_t seq_num, int32_t next_sdu_ts)
 {
-	int ret;
 	static uint8_t buf_data[CONFIG_BT_ISO_TX_MTU];
-	static bool data_initialized;
+	static size_t len_to_send = ARRAY_SIZE(buf_data);
+	bool is_first_sdu = (next_sdu_ts == -1);
 	struct net_buf *buf;
-	static size_t len_to_send = 1;
+	int ret;
 
-	if (!data_initialized) {
+	if (is_first_sdu) {
+		/* initialize mock data */
 		for (int i = 0; i < ARRAY_SIZE(buf_data); i++) {
 			buf_data[i] = (uint8_t)i;
 		}
-
-		data_initialized = true;
 	}
 
 	buf = net_buf_alloc(&tx_pool, K_FOREVER);
 	net_buf_reserve(buf, BT_ISO_CHAN_SEND_RESERVE);
-
 	net_buf_add_mem(buf, buf_data, len_to_send);
 
-	ret = bt_iso_chan_send(&iso_chan, buf, seq_num++);
+	if (unlikely(is_first_sdu)) {
+		ret = bt_iso_chan_send(&iso_chan, buf, seq_num);
+	} else {
+		ret = bt_iso_chan_send_ts(&iso_chan, buf, seq_num, next_sdu_ts);
+	}
 
 	if (ret < 0) {
 		printk("Failed to send ISO data (%d)\n", ret);
 		net_buf_unref(buf);
 	}
 
-	k_work_schedule(&iso_send_work, K_USEC(interval_us));
+	len_to_send--;
+	if (len_to_send == 0) {
+		len_to_send = ARRAY_SIZE(buf_data);
+	}
+}
 
-	len_to_send++;
-	if (len_to_send > ARRAY_SIZE(buf_data)) {
-		len_to_send = 1;
+static void iso_sent(struct bt_iso_chan *chan)
+{
+	struct bt_iso_tx_info iso_tx_info = {0};
+	int err;
+
+	err = bt_iso_chan_get_tx_sync(chan, &iso_tx_info);
+	if (err) {
+		printk("Failed obtaining timestamp\n");
+		return;
+	}
+
+	if (iso_tx_info.offset) {
+		printk("Offset should be zero for unframed PDUs\n");
+		return;
 	}
+
+	iso_send_sdu(iso_tx_info.seq_num + 1, iso_tx_info.ts + INTERVAL_US);
```

Comment:
One of the reasons why the TX was done with a timer before, was to avoid attempting to transmit again in the RX thread (which will not exist in the future, but does today). 

For this very simple sample it may be OK, but it's not a recommended approach, and thus I don't think it's the right way of triggering TX in this sample either. 

3 alternatives:
1) Restore the timer based approach (which definitely has it's own problems)
2) Submit a k_work item to send from the system workqueue
3) Add a new thread to handle TX

Reply:
The callback executes from the system work queue, not from the RX thread - see my corresponding commit comment. I think that's fine then, right?

> 3 alternatives:
> 
> 1. Restore the timer based approach (which definitely has it's own problems)

The timer approach is "better" than the previous delayed work approach in the sense that it allows for perfect syntonization on the central - if and only if one relies on implementation details of the controller. But then the phase cannot be controlled without accessing even more low-level controller implementation details plus the approach doesn't work at all on peripherals as it cannot properly account for clock drift w/o again accessing either the mentioned HCI command or accessing low-level controller implementation details that may not even be available on some implementations.

We might think about combining a timer with the HCI sync info command but that would be less spec conforming and more brittle than the approach I'm using now - see the argumentation in my response to @rugeGerritsen above. I did a LOT of testing with the timer approach - it's really hard to make this deterministic enough to work under all circumstances.

Abstractly speaking: The pure timer approach is an open control loop w/o any chance to recover from timing errors even when running for hours or days. The callback-approach or HCI approach are closed control loops which by design adjust for many small timing errors that may exist in the stack or the application and is therefore much more stable.

Plus: Only the spec-conforming approach provides us with a long-term perspective to find an implementation that no longer depends on implementation details.

> 2. Submit a k_work item to send from the system workqueue
> 3. Add a new thread to handle TX

See above, no need as the initial assumption that we're on the RX thread is incorrect IMO.